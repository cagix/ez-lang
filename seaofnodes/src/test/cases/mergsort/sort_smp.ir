START:                            [[   ]]
   7 Start   ____ ____ ____       [[    15   18   22   27   31   45    6   37  189  158   62   85  146  ]]  [  Ctrl, #TOP, int]

L6:                               [[  START      ]]
   6 sys.ary ____    7            [[    14   17   21   24   26   29   30  241    5  ]]  Ctrl
  14 len        6   15            [[     9       ]]  int
  17 $mem       6   18            [[    16   19  ]]  #BOT
  30 $rpc       6   31            [[     2       ]]  $[ALL]
  29 soff       6   15            [[    28       ]]  int
  26 src        6   27            [[    25       ]]  *[u8]
  24 doff       6   15            [[    23       ]]  int
  21 dst        6   22            [[    20    2  ]]  *[u8]
 241 #0         6                 [[    11       ]]  0

L189:                             [[  START      ]]
 189 merge   ____    7            [[   206  209  212  219  222  226  240  242   56  188  ]]  Ctrl
 206 begin    189   15            [[   205  223  ]]  int
 209 middle   189   15            [[   203  213  ]]  int
 212 end      189   15            [[   210  237  ]]  int
 240 $rpc     189   31            [[   185       ]]  $[ALL]
 226 a        189   45            [[   225  229  216  230  ]]  *[int]
 222 b        189   45            [[   221  227  ]]  *[int]
 219 $mem     189   18            [[   218  239  ]]  #BOT
  56 #1       189                 [[   236  236  ]]  1
 242 #0       189                 [[   185  236  236  ]]  0

L37:                              [[  START      ]]
  37 eq      ____    7            [[    43   44   49   52   57   12   36  ]]  Ctrl
  43 $mem      37   18            [[    41   48   54  ]]  #BOT
  44 a         37   45            [[    41       ]]  *[int]
  57 $rpc      37   31            [[    32       ]]  $[ALL]
  52 n         37   15            [[    50       ]]  int
  49 b         37   45            [[    48       ]]  *[int]
  12 #0        37                 [[    55   46  ]]  0

L62:                              [[  START      ]]
  62 copy_ar ____    7            [[    67   69   71   74   76   77   61  ]]  Ctrl
  67 begin     62   15            [[    66       ]]  int
  77 $rpc      62   31            [[    58       ]]  $[ALL]
  76 a         62   45            [[    75       ]]  *[int]
  74 b         62   45            [[    73       ]]  *[int]
  71 $mem      62   18            [[    70   72  ]]  #BOT
  69 end       62   15            [[    64       ]]  int

L85:                              [[  START      ]]
  85 main    ____    7            [[    87  138   91  134   97  108  107  106  105  104  103  102  101  100   95   99   94  120  119  118  117  116  115  114  113  112   92  111   90  132  131  130  129  128  127  126  125  124   88  123   86   84  ]]  Ctrl
  87 $mem      85   18            [[    86   97   97  ]]  #BOT
 138 $rpc      85   31            [[    78       ]]  $[ALL]
 108 st8       85  109   98 ____ ____  [[   107  ]]  #19:int
 107 st8       85  108   98 ____ ____  [[   106  ]]  #19:int
 132 st8       85  133  122 ____ ____  [[   131  ]]  #19:int
 106 st8       85  107   98 ____ ____  [[   105  ]]  #19:int
 131 st8       85  132  122 ____ ____  [[   130  ]]  #19:int
 105 st8       85  106   98 ____ ____  [[   104  ]]  #19:int
 120 st8       85  121  110 ____ ____  [[   119  ]]  #19:int
 130 st8       85  131  122 ____ ____  [[   129  ]]  #19:int
 104 st8       85  105   98 ____ ____  [[   103  ]]  #19:int
 129 st8       85  130  122 ____ ____  [[   128  ]]  #19:int
 119 st8       85  120  110 ____ ____  [[   118  ]]  #19:int
 103 st8       85  104   98 ____ ____  [[   102  ]]  #19:int
 128 st8       85  129  122 ____ ____  [[   127  ]]  #19:int
 118 st8       85  119  110 ____ ____  [[   117  ]]  #19:int
 127 st8       85  128  122 ____ ____  [[   126  ]]  #19:int
 102 st8       85  103   98 ____ ____  [[   101  ]]  #19:int
 117 st8       85  118  110 ____ ____  [[   116  ]]  #19:int
 126 st8       85  127  122 ____ ____  [[   125  ]]  #19:int
 101 st8       85  102   98 ____ ____  [[   100  ]]  #19:int
 116 st8       85  117  110 ____ ____  [[   115  ]]  #19:int
 125 st8       85  126  122 ____ ____  [[   124  ]]  #19:int
 115 st8       85  116  110 ____ ____  [[   114  ]]  #19:int
 100 st8       85  101   98 ____ ____  [[    99  ]]  #19:int
 124 st8       85  125  122 ____ ____  [[   123  ]]  #19:int
 114 st8       85  115  110 ____ ____  [[   113  ]]  #19:int
  99 st8       85  100   98 ____ ____  [[    94  ]]  #19:int
  95 st4       85   96   98 ____ ____  [[    94  ]]  #18:u32
 123 st8       85  124  122 ____ ____  [[    86  ]]  #19:int
  88 st4       85   89  122 ____ ____  [[    86  ]]  #18:u32
 113 st8       85  114  110 ____ ____  [[   112  ]]  #19:int
  94 new_ary   85   91   95   99  [[    93  110  121  ]]  [  Ctrl, *[int]!, #18:u32, #19:int]
  93 $18       94                 [[    92       ]]  #18:u32
 110 [int]     94                 [[    92  120  119  118  117  116  115  114  113  112  111   84  ]]  *[int]
 121 $19       94                 [[   120       ]]  #19:int
  86 ALLMEM    85   87 ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____   88  123  [[    84  ]]  #BOT
 112 st8       85  113  110 ____ ____  [[   111  ]]  #19:int
  84 call      85   86   98  110  134  [[    83  ]]  Ctrl
  92 st4       85   93  110 ____ ____  [[    90  ]]  #18:u32
 111 st8       85  112  110 ____ ____  [[    90  ]]  #19:int
  91 #88       85                 [[    90   94   97  ]]  88
  90 new_ary   85   91   92  111  [[    89  122  133  ]]  [  Ctrl, *[int]!, #18:u32, #19:int]
  89 $18       90                 [[    88       ]]  #18:u32
 122 [int]     90                 [[    88  132  131  130  129  128  127  126  125  124  123   81  ]]  *[int]
 133 $19       90                 [[   132       ]]  #19:int
  97 new_ary   85   91   87   87  [[    96   98  109  ]]  [  Ctrl, *[int]!, #18:u32, #19:int]
  96 $18       97                 [[    95       ]]  #18:u32
  98 [int]     97                 [[    95  108  107  106  105  104  103  102  101  100   99   84   81  ]]  *[int]
 109 $19       97                 [[   108       ]]  #19:int
 134 #10       85                 [[    84   81  ]]  10

L158:                             [[  START      ]]
 158 split_m ____    7            [[   162  163  174  175  179  184  161  159  157  ]]  Ctrl
 162 end      158   15            [[   161  177  169  166  ]]  int
 163 begin    158   15            [[   161  172  177  166  ]]  int
 174 $mem     158   18            [[   172  182  ]]  #BOT
 175 a        158   45            [[   172  169  166  ]]  *[int]
 184 $rpc     158   31            [[   154       ]]  $[ALL]
 179 b        158   45            [[   172  169  166  ]]  *[int]
 161 sub      158  162  163       [[   159       ]]  int
 159 cmp      158  161            [[   157       ]]  bool
 157 j<=      158  159            [[   156  173  ]]  [  Ctrl, Ctrl]

L146:                             [[  START      ]]
 146 merge_s ____    7            [[   147  148  149  150  153  244  145  ]]  Ctrl
 147 $mem     146   18            [[   145       ]]  #BOT
 148 a        146   45            [[   145  142  ]]  *[int]
 149 n        146   15            [[   145  142  ]]  int
 153 $rpc     146   31            [[   139       ]]  $[ALL]
 150 b        146   45            [[   145  142  ]]  *[int]
 145 call     146  147  148  244  149  150  [[   144  ]]  Ctrl
 244 #0       146                 [[   139  142  145  ]]  0

LOOP5:                            [[  L6        L8         ]]
   5 Loop    ____    6    8       [[    11   19    9    4  ]]  Ctrl
  11 Phi_i      5  241   13       [[     9   13   23   28  ]]  int
  19 Phi_$14    5   17   20       [[    16   20   25  ]]  #BOT
   9 cmp        5   11   14       [[     4       ]]  bool
   4 j<         5    9            [[     3    8  ]]  [  Ctrl, Ctrl]

LOOP188:                          [[  L189      L190       ]]
 188 Loop    ____  189  190       [[   205  213  218  223  237  187  ]]  Ctrl
 205 Phi_i    188  206  207       [[   203  208  207  225  216  ]]  int
 213 Phi_j    188  209  214       [[   210  214  215  229  230  ]]  int
 218 Phi_$19  188  219  220       [[   216  221  225  227  229  230  239  ]]  #BOT
 223 Phi_k    188  206  224       [[   221  224  227  237  ]]  int
 237 cmp      188  223  212       [[   187       ]]  bool
 187 j<       188  237            [[   186  202  ]]  [  Ctrl, Ctrl]

LOOP36:                           [[  L37       L38        ]]
  36 Loop    ____   37   38       [[    46   50   35  ]]  Ctrl
  46 Phi_i     36   12   47       [[    41   47   48   50  ]]  int
  50 cmp       36   46   52       [[    35       ]]  bool
  35 j<        36   50            [[    34   40  ]]  [  Ctrl, Ctrl]

LOOP61:                           [[  L62       L63        ]]
  61 Loop    ____   62   63       [[    66   72   64   60  ]]  Ctrl
  66 Phi_k     61   67   68       [[    64   68   73   75  ]]  int
  72 Phi_$19   61   71   73       [[    70   73   75  ]]  #BOT
  64 cmp       61   66   69       [[    60       ]]  bool
  60 j<        61   64            [[    59   63  ]]  [  Ctrl, Ctrl]

L173:                             [[  L158       ]]
 173 False    157                 [[   177  178  176  172  ]]  Ctrl
 172 call     173  174  175  163  176  179  [[   171  ]]  Ctrl
 177 add      173  162  163       [[   176       ]]  int
 178 #2       173                 [[   176       ]]  2
 176 div      173  177  178       [[   172  169  166  ]]  int

L83:                              [[  L85        ]]
  83 CallEnd   84                 [[    82  135  ]]  [  Ctrl, #BOT, 0]
 135 $mem      83                 [[    81       ]]  #BOT

L156:                             [[  L158       ]]
 156 True     157                 [[   155       ]]  Ctrl

L144:                             [[  L146       ]]
 144 CallEnd  145                 [[   143  151  ]]  [  Ctrl, #BOT, 0]
 151 $mem     144                 [[   142       ]]  #BOT

L3:                               [[  LOOP5      ]]
   3 False      4                 [[    16    2  ]]  Ctrl
  16 ALLMEM     3   17 ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____   19  [[     2  ]]  #BOT
   2 Return     3   16   21   30  [[     1       ]]  [  Ctrl, #BOT, *[u8]!]

L186:                             [[  LOOP188    ]]
 186 False    187                 [[   239  185  ]]  Ctrl
 239 ALLMEM   186  219 ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____  218  [[   185  ]]  #BOT
 185 Return   186  239  242  240  [[     1       ]]  [  Ctrl, #BOT, 0]

L8:                               [[  LOOP5      ]]
   8 True       4                 [[    28   25   23   20   13    5  ]]  Ctrl
  28 add        8   11   29       [[    25       ]]  int
  13 inc        8   11            [[    11       ]]  int
  25 ld1        8   19   26   28  [[    20       ]]  u8
  23 add        8   11   24       [[    20       ]]  int
  20 st1        8   19   21   23   25  [[    19  ]]  #14:Bot

L143:                             [[  L144       ]]
 143 $ctrl    144                 [[   142       ]]  Ctrl
 142 call     143  151  148  244  149  150  [[   141  ]]  Ctrl

L34:                              [[  LOOP36     ]]
  34 False     35                 [[   246   33  ]]  Ctrl
 246 #1        34                 [[    55       ]]  1

L40:                              [[  LOOP36     ]]
  40 True      35                 [[    48   41   39  ]]  Ctrl
  48 ld8       40   43   49   46  [[    41       ]]  int
  41 cmp8      40   43   44   46   48  [[    39  ]]  bool
  39 j==       40   41            [[    38   53  ]]  [  Ctrl, Ctrl]

L82:                              [[  L83        ]]
  82 $ctrl     83                 [[    81       ]]  Ctrl
  81 call      82  135   98  122  134  [[    80  ]]  Ctrl

L202:                             [[  LOOP188    ]]
 202 True     187                 [[   203  201  ]]  Ctrl
 203 cmp      202  205  209       [[   201       ]]  bool
 201 j<       202  203            [[   200  231  ]]  [  Ctrl, Ctrl]

L59:                              [[  LOOP61     ]]
  59 False     60                 [[    70  245   58  ]]  Ctrl
  70 ALLMEM    59   71 ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____   72  [[    58  ]]  #BOT
 245 #0        59                 [[    58       ]]  0
  58 Return    59   70  245   77  [[     1       ]]  [  Ctrl, #BOT, 0]

L63:                              [[  LOOP61     ]]
  63 True      60                 [[    75   73   68   61  ]]  Ctrl
  68 inc       63   66            [[    66       ]]  int
  75 ld8       63   72   76   66  [[    73       ]]  int
  73 st8       63   72   74   66   75  [[    72  ]]  #19:Bot

L171:                             [[  L173       ]]
 171 CallEnd  172                 [[   170  180  ]]  [  Ctrl, #BOT, 0]
 180 $mem     171                 [[   169       ]]  #BOT

L141:                             [[  L143       ]]
 141 CallEnd  142                 [[   140  152  ]]  [  Ctrl, #BOT, 0]
 152 $mem     141                 [[   139       ]]  #BOT

L80:                              [[  L82        ]]
  80 CallEnd   81                 [[    79  136  137  ]]  [  Ctrl, #BOT, bool]
 136 $mem      80                 [[    78       ]]  #BOT
 137 #2        80                 [[    78       ]]  bool

L170:                             [[  L171       ]]
 170 $ctrl    171                 [[   169       ]]  Ctrl
 169 call     170  180  175  176  162  179  [[   168  ]]  Ctrl

L231:                             [[  L202       ]]
 231 False    201                 [[   195       ]]  Ctrl

L200:                             [[  L202       ]]
 200 True     201                 [[   210  199  ]]  Ctrl
 210 cmp      200  212  213       [[   199       ]]  bool
 199 j<=      200  210            [[   198  232  ]]  [  Ctrl, Ctrl]

L38:                              [[  L40        ]]
  38 True      39                 [[    47   36  ]]  Ctrl
  47 inc       38   46            [[    46       ]]  int

L53:                              [[  L40        ]]
  53 False     39                 [[    33       ]]  Ctrl

L140:                             [[  L141       ]]
 140 $ctrl    141                 [[   139       ]]  Ctrl
 139 Return   140  152  244  153  [[     1       ]]  [  Ctrl, #BOT, 0]

L33:                              [[  L34       L53        ]]
  33 Region  ____   34   53       [[    55   54   32  ]]  Ctrl
  55 Phi_res   33  246   12       [[    32       ]]  bool
  54 ALLMEM    33   43 ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____ ____  [[    32  ]]  #BOT
  32 Return    33   54   55   57  [[     1       ]]  [  Ctrl, #BOT, bool]

L79:                              [[  L80        ]]
  79 $ctrl     80                 [[    78       ]]  Ctrl
  78 Return    79  136  137  138  [[     1       ]]  [  Ctrl, #BOT, bool]

L198:                             [[  L200       ]]
 198 False    199                 [[   230  216  197  ]]  Ctrl
 230 ld8      198  218  226  213  [[   216       ]]  int
 216 cmp8     198  218  226  205  230  [[   197  ]]  bool
 197 j<=      198  216            [[   196  233  ]]  [  Ctrl, Ctrl]

L168:                             [[  L170       ]]
 168 CallEnd  169                 [[   167  181  ]]  [  Ctrl, #BOT, 0]
 181 $mem     168                 [[   166       ]]  #BOT

L232:                             [[  L200       ]]
 232 True     199                 [[   195       ]]  Ctrl

L167:                             [[  L168       ]]
 167 $ctrl    168                 [[   166       ]]  Ctrl
 166 call     167  181  179  163  176  162  175  [[   165  ]]  Ctrl

L233:                             [[  L198       ]]
 233 False    197                 [[   195       ]]  Ctrl

L196:                             [[  L198       ]]
 196 True     197                 [[   195       ]]  Ctrl

L165:                             [[  L167       ]]
 165 CallEnd  166                 [[   164  183  ]]  [  Ctrl, #BOT, 0]
 183 $mem     165                 [[   182       ]]  #BOT

L195:                             [[  L196      L231      L232      L233       ]]
 195 Region  ____  196  231  232  233  [[   236  234  194  ]]  Ctrl
 236 Phi_con  195   56  242   56  242  [[   234  ]]  bool
 234 test     195  236            [[   194       ]]
 194 j!=      195  234            [[   191  228  ]]  [  Ctrl, Ctrl]

L164:                             [[  L165       ]]
 164 $ctrl    165                 [[   155       ]]  Ctrl

L191:                             [[  L195       ]]
 191 True     194                 [[   225  221  208  190  ]]  Ctrl
 208 inc      191  205            [[   207       ]]  int
 225 ld8      191  218  226  205  [[   221       ]]  int
 221 st8      191  218  222  223  225  [[   220  ]]  #19:Bot

L155:                             [[  L156      L164       ]]
 155 Region  ____  156  164       [[   182  243  154  ]]  Ctrl
 182 Phi_$me  155  174  183       [[   154       ]]  #BOT
 243 #0       155                 [[   154       ]]  0
 154 Return   155  182  243  184  [[     1       ]]  [  Ctrl, #BOT, 0]

L228:                             [[  L195       ]]
 228 False    194                 [[   229  227  215  190  ]]  Ctrl
 215 inc      228  213            [[   214       ]]  int
 229 ld8      228  218  226  213  [[   227       ]]  int
 227 st8      228  218  222  223  229  [[   220  ]]  #19:Bot

L190:                             [[  L191      L228       ]]
 190 Region  ____  191  228       [[   207  214  220  224  188  ]]  Ctrl
 207 Phi_i    190  208  205       [[   205       ]]  int
 214 Phi_j    190  213  215       [[   213       ]]  int
 220 Phi_$19  190  221  227       [[   218       ]]  #19:Bot
 224 inc      190  223            [[   223       ]]  int

L1:                               [[  L79       L140      L155      L186       ]]
   1 Stop       2   32   58   78  139  154  185  [[   ]]  Bot