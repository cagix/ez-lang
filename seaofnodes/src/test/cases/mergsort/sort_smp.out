---split_merge { *![int] int int *![int] -> 0 #0}---------------------------
0000 4883EC38    subi  rsp -= #56
                 a:r9,b:rcx,begin:rdx,end:r8
0004 4889542428  mov   [rsp+40] = rdx                 // def/loop #1
0009 4C89442458  mov   [rsp+88] = r8                  // def/loop #0
000E 48894C2450  mov   [rsp+80] = rcx                 // def/loop #0
0013 4C894C2448  mov   [rsp+72] = r9                  // def/loop #0
0018 4C8B442458  mov   r8 = [rsp+88]                  // use/loop/use #0
001D 4C8B542428  mov   r10 = [rsp+40]                 // use/loop/use #1
0022 4D2BC2      sub   r8 -= r10
0025 4983F801    cmp   r8, #1
0029 7E6F        j<=   L90                            // L98
L98:
002B 488B442458  mov   rax = [rsp+88]                 // use/loop/use #0
0030 488B542428  mov   rdx = [rsp+40]                 // use/loop/use #1
0035 4803C2      add   rax += rdx
0038 B902000000  ldi   rcx = #2
003D 489948F7F9  div   rax = rax / rcx                // kill rdx
0042 4889442440  mov   [rsp+64] = rax                 // def/loop #0
0047 488B4C2448  mov   rcx = [rsp+72]                 // use/loop/use #0
004C 4C8B4C2450  mov   r9 = [rsp+80]                  // use/loop/use #0
0051 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #0
0056 488B542428  mov   rdx = [rsp+40]                 // use/loop/use #1
005B E8A0FFFFFF  call  split_merge  rcx  rdx  r8  r9
0060 488B4C2448  mov   rcx = [rsp+72]                 // use/loop/use #0
0065 4C8B4C2450  mov   r9 = [rsp+80]                  // use/loop/use #0
006A 4C8B442458  mov   r8 = [rsp+88]                  // use/loop/use #0
006F 488B542440  mov   rdx = [rsp+64]                 // use/loop/use #0
0074 E887FFFFFF  call  split_merge  rcx  rdx  r8  r9
0079 FF7424488F  mov   [rsp+32] = [rsp+72]            // use/loop/use #0
007E 442420
0081 488B4C2450  mov   rcx = [rsp+80]                 // use/loop/use #0
0086 4C8B4C2458  mov   r9 = [rsp+88]                  // use/loop/use #0
008B 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #0
0090 488B542428  mov   rdx = [rsp+40]                 // use/loop/use #1
0095 E836000000  call  merge  rcx  rdx  r8  r9  [stk#42]
L91:
L90:
009A 33C0        xor   rax,rax
009C 4883C438C3  addi  rsp += #56
00A1             ret
---{ *![int] int int *![int] -> 0 #0}---------------------------

---copy_array { *![int] int int *![int] -> 0 #1}---------------------------
                 begin:rdx,end:r8,b:r9,a:rcx
LOOP69:
                 k:rdx
00B0 493BD0      cmp   rdx, r8
00B3 7D10        j>=   L67                            // L71
L71:
00B5 488B44D108  ld8   rax,[rcx+rdx*8+8]
00BA 498944D108  st8   [r9+rdx*8+8],rax
00BF 4883C201    inc   rdx += #1
00C3 EBEB        jmp   LOOP69
L425:
L67:
00C5 33C0        xor   rax,rax
00C7 C3          ret
---{ *![int] int int *![int] -> 0 #1}---------------------------

---merge { *![int] int int int *![int] -> 0 #0}---------------------------
                 begin:rdx,middle:r8,end:r9,b:rcx,a:[rsp+40]
00D0 66480F6EC3  mov   xmm0 = rbx                     // def/loop #2
00D5 488B5C2428  mov   rbx = [rsp+40]                 // def/empty1 #0
00DA 4C8BDA      mov   r11 = rdx                      // use/self/phi #1
00DD 4D8BD0      mov   r10 = r8                       // use/self/use #1
LOOP120:
                 i:r11,j:r10,k:rdx
00E0 493BD1      cmp   rdx, r9
00E3 7D57        j>=   L118                           // L134
L134:
00E5 4D3BD8      cmp   r11, r8
00E8 7C04        j<    L132                           // L163
L163:
00EA 33C0        xor   rax,rax
00EC EB24        jmp   L127
L424:
L132:
00EE 4D3BCA      cmp   r9, r10
00F1 7E1A        j<=   L164                           // L130
L130:
00F3 488BC3      mov   rax = rbx                      // use/empty1 #0
00F6 4A8B44D008  ld8   rax,[rax+r10*8+8]
00FB 4A3944DB08  cmp8  rax, [rbx+r11*8+8]
0100 7E04        j<=   L128                           // L165
L165:
0102 33C0        xor   rax,rax
0104 EB0C        jmp   L127
L423:
L128:
0106 B801000000  ldi   rax = #1
010B EB05        jmp   L127
L422:
L164:
010D B801000000  ldi   rax = #1
L127:
                 cond:rax
0112 4883F800    test  rax
0116 7510        j!=   L123                           // L160
L160:
0118 4A8B44D308  ld8   rax,[rbx+r10*8+8]
011D 488944D108  st8   [rcx+rdx*8+8],rax
0122 4983C201    inc   r10 += #1
0126 EB0E        jmp   L122
L421:
L123:
0128 4A8B44DB08  ld8   rax,[rbx+r11*8+8]
012D 488944D108  st8   [rcx+rdx*8+8],rax
0132 4983C301    inc   r11 += #1
L122:
                 i:r11,j:r10
0136 4883C201    inc   rdx += #1
013A EBA4        jmp   LOOP120
L420:
L118:
013C 66480F7EC3  mov   rbx = xmm0                     // use/loop/use #2
0141 33C0        xor   rax,rax
0143 C3          ret
---{ *![int] int int int *![int] -> 0 #0}---------------------------

---eq { *![int] *![int] int -> bool #1}---------------------------
                 a:rcx,b:rdx,n:r8
0150 33C0        xor   rax,rax
LOOP189:
                 i:rax
0152 493BC0      cmp   rax, r8
0155 7D16        j>=   L187                           // L193
L193:
0157 4C8B4CC208  ld8   r9,[rdx+rax*8+8]
015C 4C394CC108  cmp8  r9, [rcx+rax*8+8]
0161 7506        j!=   L205                           // L191
L191:
0163 4883C001    inc   rax += #1
0167 EBE9        jmp   LOOP189
L419:
L205:
0169 33C0        xor   rax,rax
016B EB05        jmp   L186
L418:
L187:
016D B801000000  ldi   rax = #1
L186:
                 result:rax
0172 C3          ret
---{ *![int] *![int] int -> bool #1}---------------------------

---main { -> bool #0}---------------------------
0180 4883EC28    subi  rsp -= #40
0184 48895C2420  mov   [rsp+32] = rbx                 // def/loop #2
0189 BA58000000  ldi   rdx = #88
018E B901000000  alloc ldi   rcx = #1
0193 E800000000  call  #calloc
0198 4889442418  mov   [rsp+24] = rax                 // def/loop #0
019D 4C8B4C2418  mov   r9 = [rsp+24]                  // use/loop/use #0
01A2 49C741080A  st8   [r9+8],#10
01A7 000000
01AA 4C8B442418  mov   r8 = [rsp+24]                  // use/loop/use #0
01AF 49C7401009  st8   [r8+16],#9
01B4 000000
01B7 488B542418  mov   rdx = [rsp+24]                 // use/loop/use #0
01BC 48C7421808  st8   [rdx+24],#8
01C1 000000
01C4 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
01C9 48C7412007  st8   [rcx+32],#7
01CE 000000
01D1 488B442418  mov   rax = [rsp+24]                 // use/loop/use #0
01D6 48C7402806  st8   [rax+40],#6
01DB 000000
01DE 4C8B5C2418  mov   r11 = [rsp+24]                 // use/loop/use #0
01E3 49C7433005  st8   [r11+48],#5
01E8 000000
01EB 4C8B542418  mov   r10 = [rsp+24]                 // use/loop/use #0
01F0 49C7423804  st8   [r10+56],#4
01F5 000000
01F8 4C8B4C2418  mov   r9 = [rsp+24]                  // use/loop/use #0
01FD 49C7414003  st8   [r9+64],#3
0202 000000
0205 4C8B442418  mov   r8 = [rsp+24]                  // use/loop/use #0
020A 49C7404802  st8   [r8+72],#2
020F 000000
0212 488B542418  mov   rdx = [rsp+24]                 // use/loop/use #0
0217 C7020A0000  st4   [rdx],#10
021C 00
021D 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
0222 48C7415001  st8   [rcx+80],#1
0227 000000
022A BA58000000  ldi   rdx = #88
022F B901000000  alloc ldi   rcx = #1
0234 E800000000  call  #calloc
0239 4889442410  mov   [rsp+16] = rax                 // def/empty1 #1
023E C7000A0000  st4   [rax],#10
0243 00
0244 BA58000000  ldi   rdx = #88
0249 B901000000  alloc ldi   rcx = #1
024E E800000000  call  #calloc
0253 488BD8      mov   rbx = rax                      // def/empty1 #0
0256 48C7430801  st8   [rbx+8],#1
025B 000000
025E 48C7431002  st8   [rbx+16],#2
0263 000000
0266 48C7431803  st8   [rbx+24],#3
026B 000000
026E 48C7432004  st8   [rbx+32],#4
0273 000000
0276 48C7432805  st8   [rbx+40],#5
027B 000000
027E 48C7433006  st8   [rbx+48],#6
0283 000000
0286 48C7433807  st8   [rbx+56],#7
028B 000000
028E 48C7434008  st8   [rbx+64],#8
0293 000000
0296 48C7434809  st8   [rbx+72],#9
029B 000000
029E C7030A0000  st4   [rbx],#10
02A3 00
02A4 48C743500A  st8   [rbx+80],#10
02A9 000000
02AC 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
02B1 488B542410  mov   rdx = [rsp+16]                 // use/empty1 #0
02B6 41B80A0000  ldi   r8 = #10
02BB 00
02BC E81F000000  call  merge_sort  rcx  rdx  r8
02C1 488B4C2418  mov   rcx = [rsp+24]                 // use/loop/use #0
02C6 488BD3      mov   rdx = rbx                      // use/empty1 #0
02C9 41B80A0000  ldi   r8 = #10
02CE 00
02CF E87CFEFFFF  call  eq  rcx  rdx  r8
02D4 488B5C2420  mov   rbx = [rsp+32]                 // use/loop/use #2
02D9 4883C428C3  addi  rsp += #40
02DE             ret
---{ -> bool #0}---------------------------

---merge_sort { *![int] *![int] int -> 0 #0}---------------------------
02E0 4883EC28    subi  rsp -= #40
                 a:rcx,n:r8,b:rdx
02E4 4C89442440  mov   [rsp+64] = r8                  // def/loop #1
02E9 48894C2438  mov   [rsp+56] = rcx                 // def/loop #1
02EE 4889542430  mov   [rsp+48] = rdx                 // def/loop #0
02F3 4C8B4C2430  mov   r9 = [rsp+48]                  // use/loop/use #0
02F8 33D2        xor   rdx,rdx
02FA 488B4C2438  mov   rcx = [rsp+56]                 // use/loop/use #1
02FF 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #1
0304 E8A7FDFFFF  call  copy_array  rcx  rdx  r8  r9
0309 4C8B4C2430  mov   r9 = [rsp+48]                  // use/loop/use #0
030E 33D2        xor   rdx,rdx
0310 488B4C2438  mov   rcx = [rsp+56]                 // use/loop/use #1
0315 4C8B442440  mov   r8 = [rsp+64]                  // use/loop/use #1
031A E8E1FCFFFF  call  split_merge  rcx  rdx  r8  r9
031F 33C0        xor   rax,rax
0321 4883C428C3  addi  rsp += #40
0326             ret
---{ *![int] *![int] int -> 0 #0}---------------------------
